### PR TITLE
Fix Server volume and mute after reboot

### LINKED ISSUE #3105: Server's volume SynthDef 'volumeAmpControl2' not found after reboot
The server's volume control breaks after reboot if the volume was previously changed. There was some discussion in #2918 . A similar issue was apparently fixed (see #393).

### Reproducer
```supercollider
s.boot;

// set volume to -20 db
s.volume = 0.1.ampdb;

// make sound 
{SinOsc.ar}.play;

// quit and reboot server
s.quit;
s.boot;

// notice the error message: 
// *** ERROR: SynthDef volumeAmpControl2 not found
// FAILURE IN SERVER /s_new SynthDef not found
// the indicator on the bottom right says -20db

// setting volume gives error
s.volume = 0.1.ampdb;

// make sound (not-20db but 0db) 
{SinOsc.ar}.play;

// Cmd-Period

// make sound again (now -20db) 
{SinOsc.ar}.play;
```

### PR DESCRIPTION
**Pull request by patrickdupuis**
This PR is an attempt to fix #3105 & #3110.
### Description

The problem is that, on reboot, ServerTree calls its updateFunc (it was added the first time the server booted) before ServerBoot's initFunc is done sending the SynthDef. This causes the "SynthDef not found" error.
### Changes proposed

The fix removes the updateFunc from ServerTree when the server quits. That way, on reboot, ServerTree has nothing to call. When initFunc is finished sending the SynthDef, updateFunc gets added to ServerTree.

Also, this.updateSynth needs to be called at the end of sendSynthDef so that the volume synth gets created, if necessary.

This change also makes Server.quit call Volume.freeSynth.

### Result
The Server's volume and mute status persist across reboots. No more SynthDef not found errors posted after reboot.

### Test
```supercollider
s.volume.volume = -36;
s.boot;
s.queryAllNodes;
/* should see this (if you have 2 outputs)
NODE TREE Group 0
   1 group
   1000 volumeAmpControl2
*/

s.quit;
s.boot;
s.queryAllNodes;
/* should see this (if you have 2 outputs)
NODE TREE Group 0
   1 group
   1000 volumeAmpControl2
*/
a = {SinOsc.ar}.play; // should play at -36db 
a.free;

s.mute;
s.quit;
s.boot;
a = {SinOsc.ar}.play; // no sound, server muted 
s.unmute; // should play at -36db
a.free;

s.quit;
s.options.numOutputBusChannels = 4;
s.boot;
s.queryAllNodes;
/* should see:
NODE TREE Group 0
   1 group
   1000 volumeAmpControl4
*/
```

### CODE REVIEW

**Comment by patrickdupuis** [comment]
I messed up my first attempt at this PR. This is a redo.

**Comment by patrickdupuis** [comment]
Oops... I should just stop now.

**Comment by patrickdupuis** [comment]
Please review :)

**Comment by brianlheim** [comment]
would it be possible to add a regression test for this?

File: SCClassLibrary/Common/Control/Volume.sc [hunk 1]
```diff
@@ -22,19 +22,20 @@ Volume {
...
+					In.ar(bus, synthNumChans) * Lag.kr(volumeAmp, volumeLag)
+				);
 			}).send(server);
 
 			server.sync;
 
+			this.updateSynth;
+
 			if(updateFunc.isNil) {
 				ServerTree.add(updateFunc = {
```
Comment (snappizz) [inline SCClassLibrary/Common/Control/Volume.sc]: while we're at it... this use of an assignment as an expression makes me cringe, could you move it out to `updateFunc = { }; ServerTree.add(updateFunc)`? i'm warming up to brian's policy of cleaning up adjacent code when working on things.

**Comment by telephon** [comment]
please do go ahead and change that. I don't think an assignment in an argument should make you cringe, though. It can be a bit obscure. But it is a clear consequence of the fact that every expression returns a value.

but I'm fine with a change of course if it helps readability

File: SCClassLibrary/Common/Control/Volume.sc (hunk 1 above)
Comment (patrickdupuis) [inline SCClassLibrary/Common/Control/Volume.sc]: Will do!

**Comment by patrickdupuis** [comment]
I can write a unit test for this in a few days.

**Comment by patrickdupuis** [comment]
Hmmm... with my last commit I get a warning while booting. Everthing seems to work fine though.
```
SuperCollider 3 server ready.
JackDriver: max output latency 21.3 ms
WARNING: Server internal not running, could not send SynthDef.
Shared memory server interface initialized
Receiving notification messages from server 'localhost'
```
Looks like `SynthDef.send` complains that the server isn't running. I don't think this is caused by a mistake on my part. Rather, it's an issue that has become apparent with my last commit.

**Comment by patrickdupuis** [comment]
I made the server sync before creating the SynthDef. Maybe this is a little too hacky? I'm open to improvements.

File: SCClassLibrary/Common/Control/Volume.sc [hunk 2]
```diff
@@ -34,14 +35,14 @@ Volume {
 
 			server.sync;
 
-			this.updateSynth;
+			updateFunc = {
+				ampSynth = nil;
+				if(persist) { this.updateSynth }
+			};
 
-			if(updateFunc.isNil) {
```
Comment (telephon) [inline SCClassLibrary/Common/Control/Volume.sc]: I can't test it here, but could you make sure that `ServerTree` doesn't accumulate these functions?

Comment (patrickdupuis) [reply SCClassLibrary/Common/Control/Volume.sc]: I don't see any accumulation happening.

**Comment by patrickdupuis** [comment]
No, you can change the volume and mute the server as usual. The server volume mechanism still functions as before, except for the missing SynthDef errors. What gives you the impression that the server will boot when the volume changes?

File: testsuite/classlibrary/TestVolume.sc [hunk 3]
```diff
@@ -1,24 +1,49 @@
 TestVolume : UnitTest {
 
-	test_setVolume {
-		var server, volume, level, temp;
+	var server = Server.default;
+	var volume = Server.default.volume;
+	var level = Server.default.volume.volume;
+	var temp;
```
Comment (brianlheim) [inline testsuite/classlibrary/TestVolume.sc]: seems to me that a temp variable should not be declared at class scope, and probably deserves a better name

Comment (patrickdupuis) [reply testsuite/classlibrary/TestVolume.sc]: you're right. I'll fix that right away.

**Comment by patrickdupuis** [comment]
I've added a UnitTest to this PR. Please have a look.

In order to make the tests, I've had to modify Volume.sc slightly. Volume's `ampSynth` is now a getter. I've also made the local variable `synthNumChans` into a global variable called `numOutputChannels`. It's used in the SynthDef creation and can be used to get the number of channels of the ampSynth. I would have called it `numChannels` but that is already taken. To me, `numOutputChannels` is a better variable name. I can change it to `synthNumChans` if desired.

File: testsuite/classlibrary/TestVolume.sc [hunk 4]
```diff
@@ -0,0 +1,40 @@
+TestVolume : UnitTest {
+
+	var server = Server.default;
```
Comment (LFSaw) [inline testsuite/classlibrary/TestVolume.sc]: this creates a parse error. Did you test the UnitTest?

Comment (LFSaw) [reply testsuite/classlibrary/TestVolume.sc]: I started to look into this and got this far (see below). However, I got 
```
PASS:a TestVolume:test_setVolume - initial volume is 0 db
*** ERROR: SynthDef volumeAmpControl2 not found
FAILURE IN SERVER /s_new SynthDef not found
PASS:a TestVolume:test_setVolume - Server volume synth exists
FAILURE IN SERVER /n_set Node 1000 not found
FAILURE IN SERVER /s_get Node 1000 not found
ERROR: Message '-' not understood.
```

now I need to sleep... so leave this to others...

```
TestVolume : UnitTest {

	test_setVolume {
		var server = Server.default;
		var serverVolume = server.volume.volume;
		var ampSynthVolume;

		this.bootServer;

		this.assert(serverVolume == 0, "initial volume is 0 db");

		server.volume.volume = -36;
		server.sync;
		this.assert(server.volume.ampSynth.notNil, "Server volume synth exists");

		server.volume.ampSynth.get(\volumeAmp, { |level| ampSynthVolume = level });
		server.sync;
		this.assertFloatEquals(ampSynthVolume, serverVolume, "volume level correctly set");

		server.volume.reset;
		server.sync;
		server.quit;
	}

// [...]
}
```

**Comment by patrickdupuis** [comment]
@LFSaw thanks for looking at this. The tests should work now 👍

**Comment by LFSaw** [comment]
I confirmed with running the UnitTests locally.

**Comment by brianlheim** [comment]
thanks for that @LFSaw :)

**Pull request merged**